quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/infinispan/jgroups/impl/JpaJGroupsTlsConfigurator.java

```diff
@@ -0,0 +1,116 @@
+    private static final String ALIAS = "jgroups";
+    public static final JpaJGroupsTlsConfigurator INSTANCE = new JpaJGroupsTlsConfigurator();
+
+    @Override
+    public boolean requiresKeycloakSession() {
+        return true;
+    }
+
+    @Override
+    SocketFactory createSocketFactory(KeycloakSession session) {
+        try (var innerSession = session.getKeycloakSessionFactory().create()) {
+            innerSession.getTransactionManager().begin();
+            var storage = innerSession.getProvider(CertificateStorageProvider.class);
+            var data = storage.loadOrCreate(CERTIFICATE_ID, JpaJGroupsTlsConfigurator::generateSelfSignedCertificate);
+            var km = createKeyManager(data.keyPair(), data.certificate());
+            var tm = createTrustManager(data.certificate());
+            var sslContext = SSLContext.getInstance("TLS");
+            sslContext.init(new KeyManager[]{km}, new TrustManager[]{tm}, null);
+            return createFromContext(sslContext);
+        } catch (IOException | GeneralSecurityException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private X509ExtendedKeyManager createKeyManager(KeyPair keyPair, X509Certificate certificate) throws GeneralSecurityException, IOException {
+        var ks = CryptoIntegration.getProvider().getKeyStore(KeystoreUtil.KeystoreFormat.JKS);
+        ks.load(null, null);
+        ks.setKeyEntry(ALIAS, keyPair.getPrivate(), KEY_PASSWORD, new java.security.cert.Certificate[]{certificate});
+        var kmf = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm());
+        kmf.init(ks, KEY_PASSWORD);
+        for (KeyManager km : kmf.getKeyManagers()) {
+            if (km instanceof X509ExtendedKeyManager) {
+                return (X509ExtendedKeyManager) km;
+            }
+        }
+        throw new GeneralSecurityException("Could not obtain an X509ExtendedKeyManager");
+    }
+
+    private X509ExtendedTrustManager createTrustManager(X509Certificate certificate) throws GeneralSecurityException, IOException {
+        var ks = CryptoIntegration.getProvider().getKeyStore(KeystoreUtil.KeystoreFormat.JKS);
+        ks.load(null, null);
+        ks.setCertificateEntry(ALIAS, certificate);
+        var tmf = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
+        tmf.init(ks);
+        for (TrustManager tm : tmf.getTrustManagers()) {
+            if (tm instanceof X509ExtendedTrustManager) {
+                return (X509ExtendedTrustManager) tm;
+            }
+        }
+        throw new GeneralSecurityException("Could not obtain an X509TrustManager");
+    }
+
+    private static CertificateDetail generateSelfSignedCertificate() {
+        var keyPair = KeyUtils.generateRsaKeyPair(2048);
+        var certificate = CertificateUtils.generateV1SelfSignedCertificate(keyPair, "jgroups");
+        return new CertificateDetail(keyPair, certificate);
+    }
```

Comment:
Let me know if this is safe or not. The idea is to have a self-signed certificate used to configure TLS between the Keycloak servers.

Reply:
Should be safe enough, we use it in other places as well.